test/integration/src/test/java/io/pravega/test/integration/TruncateTest.java

```diff
@@ -0,0 +1,113 @@
+        this.serviceBuilder.close();
+    }
+    
+    @Test
+    public void appendThroughStreamingClient() throws InterruptedException, ExecutionException, TimeoutException, TruncatedDataException, ReinitializationRequiredException {
+        String endpoint = "localhost";
+        String streamName = "abc";
+        int port = TestUtils.getAvailableListenPort();
+        String testString = "Hello world\n";
+        StreamSegmentStore store = this.serviceBuilder.createStreamSegmentService();
+        @Cleanup
+        PravegaConnectionListener server = new PravegaConnectionListener(false, port, store);
+        server.startListening();
+        @Cleanup
+        MockStreamManager streamManager = new MockStreamManager(SCOPE, endpoint, port);
+        @Cleanup
+        MockClientFactory clientFactory = streamManager.getClientFactory();
+        streamManager.createScope(SCOPE);
+        streamManager.createStream(SCOPE, streamName, null);
+        Serializer<String> serializer = new JavaSerializer<>();
+        @Cleanup
+        EventStreamWriter<String> producer = clientFactory.createEventWriter(streamName, serializer, EventWriterConfig.builder().build());
+        Future<Void> ack = producer.writeEvent(testString);
+        ack.get(5, TimeUnit.SECONDS);
+        
+        SegmentMetadataClientFactory metadataClientFactory = new SegmentMetadataClientFactoryImpl(new MockController(endpoint, port, streamManager.getConnectionFactory()), streamManager.getConnectionFactory());
+        SegmentMetadataClient metadataClient = metadataClientFactory.createSegmentMetadataClient(new Segment(SCOPE, streamName, 0), "");
+        assertEquals(0, metadataClient.getSegmentInfo().getStartingOffset());
+        long writeOffset = metadataClient.getSegmentInfo().getWriteOffset();
+        assertEquals(writeOffset, metadataClient.fetchCurrentSegmentLength());
+        assertTrue(metadataClient.getSegmentInfo().getWriteOffset() > testString.length());
+        metadataClient.truncateSegment(writeOffset);
+        assertEquals(writeOffset, metadataClient.getSegmentInfo().getStartingOffset());
+        assertEquals(writeOffset, metadataClient.getSegmentInfo().getWriteOffset());
+        assertEquals(writeOffset, metadataClient.fetchCurrentSegmentLength());
+        
+        ack = producer.writeEvent(testString);
+        ack.get(5, TimeUnit.SECONDS);
+        
+        streamManager.createReaderGroup("ReaderGroup", ReaderGroupConfig.builder().disableAutomaticCheckpoints().stream(new StreamImpl(SCOPE, streamName)).build());
+        @Cleanup
+        EventStreamReader<String> reader = clientFactory.createReader("reader", "ReaderGroup", serializer, ReaderConfig.builder().build());
+        AssertExtensions.assertThrows(TruncatedDataException.class, () -> reader.readNextEvent(2000));
+        EventRead<String> event = reader.readNextEvent(2000);
+        assertEquals(testString, event.getEvent());
+        event = reader.readNextEvent(100);
```

Comment:
This is not actually testing the NPE path as the truncation is happening at the write offset, leaving the segment empty. Of course, it relies on the truncation doing the right thing, but I think that to test that we don't throw in the presence of an imprecise truncation, we would need to leave at least one byte in the segment.

Reply:
The test testStreamTruncatedWithPartialEvent I added above covers that.